rosbag2_transport/src/rosbag2_transport/recorder.cpp

```diff
@@ -83,4 +75,65 @@ Recorder::create_subscription(
   return subscription;
 }
 
+std::future<void> Recorder::launch_topics_discovery(
+  std::chrono::milliseconds topic_polling_interval,
+  bool is_discovery_disabled,
+  const std::vector<std::string> & topics_to_record)
+{
+  auto subscribe_to_topics =
+    [this, topics_to_record, topic_polling_interval, is_discovery_disabled] {
+      do {
+        auto all_topics_and_types_to_subscribe = topics_to_record.empty() ?
```

Comment:
I think this is pretty complicated for what it is.
wouldn't it be easier to have this function called once as a blocking call to start the recording and then asynchronously call `node_->get_topics_with_types` and subscribe to the not-yet-subscribed once? I think this makes it easier to follow what's happening here.

By extracting the new topics, the call to `is_every_topic_subscribed` is then not necessary anymore, because we know which topics to subscribe.

Reply:
> By extracting the new topics, the call to `is_every_topic_subscribed` is then not necessary anymore, because we know which topics to subscribe.

I don't understand. In the `--all` case we do not know what topics to expect.